### DISCUSSION TITLE
Question about modification of `PIController`

### DISCUSSION THREAD

**Discussion by xiaoheimai95** [CURENT/andes#548] [General]
Dear Dr. Cui and Jinning,
I modified the model `PIController` in order to add an external parameter `utd` and `utq` to inffect the voltage of VSG. The main modification of `PIController` is as follows:
```
class PIControllerwithut(Block):
    """
    Proportional Integral Controller with external control variable ut.
    """

    def __init__(self, u, kp, ki, ut, ref=0.0, x0=0.0, name=None, tex_name=None, info=None):
        Block.__init__(self, name=name, tex_name=tex_name, info=info)

        self.u = dummify(u)
        self.kp = dummify(kp)
        self.ki = dummify(ki)
        self.ut = dummify(ut)
        self.ref = dummify(ref)
        self.x0 = dummify(x0)

        self.xi = State(info="Integrator output", tex_name='xi')
        self.y = Algeb(info="PI output", tex_name='y')

        self.vars = OrderedDict([('xi', self.xi), ('y', self.y)])

    def define(self):
        r"""
        Define equations for the PI Controller with external control variable ut.

        Notes
        -----
        One state variable ``xi`` and one algebraic variable ``y`` are added.

        Equations implemented are

        .. math ::
            \dot{x_i} &= k_i * (u + ut - ref) \\
            y &= x_i + k_p * (u + ut - ref)
        """

        self.xi.v_str = f'{self.x0.name}'
        self.xi.e_str = f'{self.ki.name} * ({self.u.name} + {self.ut.name} - {self.ref.name})'

        self.y.v_str = f'{self.kp.name} * ({self.u.name} + {self.ut.name} - {self.ref.name}) + {self.x0.name}'
        self.y.e_str = f'{self.kp.name} * ({self.u.name} + {self.ut.name} - {self.ref.name}) + ' \
                       f'{self.name}_xi - {self.name}_y'
```
However, when I test the effect of the modification, it came up with the problem of initialization failure on the case of SMIB.txt, I will upload the modified SMIB.txt and init_info.txt later.
So is it possible for you to help me figure out this problem?
[info.txt](https://github.com/user-attachments/files/15896037/info.txt)
[SMIB.xlsx](https://github.com/user-attachments/files/15896038/SMIB.xlsx)
![error](https://github.com/CURENT/andes/assets/71546641/71c97c8e-3712-4e01-9560-61e03ffec112)

**Comment by jinningwang** [comment]
Check the v_str and e_str of xi. As discussed in documentation, v_str is used to initialize the equation.
Please ensure the v_str and e_str are consistent.

**Reply by xiaoheimai95** [reply]
I recheck the doc of REGCV3, the e_str and v_str of xi is as bellow:
```
Initialization Equations

  Name   |    Type    |              Initial Value
---------+------------+-----------------------------------------
 dw      | State      | 0
 omega   | State      | u
 delta   | State      | a
 PIvd_xi | State      | (Id0)
 PIvq_xi | State      | (Iq0)
```

```
Differential Equations

  Name   | Type  | RHS of Equation "T x' = f(x, y)" | T (LHS)
---------+-------+----------------------------------+--------
 dw      | State | Pref2 - Pe - D * dw              | M
 omega   | State | 1 + dw - omega                   |
 delta   | State | 2 * pi * fn * dw                 |
 PIvd_xi | State | Kivd * ((vd-vref2)  + utd - 0.0) |
 PIvq_xi | State | Kivq * ((vq)  + utq - 0.0)       |
 LGId_y  | State | (1 * PIvd_y - 1 * LGId_y)        | Tid
 LGIq_y  | State | (1 * PIvq_y - 1 * LGIq_y)        | Tiq
```
Would you please tell me how can I modified the v_str of xi?

**Reply by jinningwang** [reply]
This can be a little bit tricky. Let me think about how to explain it in a user-friendly manner.

**Reply by xiaoheimai95** [reply]
Thanks a lot, looking forward to your further replay.

**Reply by jinningwang** [reply]
It is a little bit weird to me that the Eqn. mismatch seems to be 0. Is it the numerical issues that they are not actually 0?

The documentation discussed the value and equation strings: [Development - Variables](https://docs.andes.app/en/latest/modeling/variables.html#value-and-equation-strings)

Further, in the [Example - TGOV1](https://docs.andes.app/en/latest/modeling/examples.html#tgov1), there is a definition of algebraic variable `pd`.
```python
  self.pd = Algeb(info='Droop output',
              v_str='tm0',
              e_str='(wd+pref)/R-pd')
```
Given its `v_str` and `e_str`, it should satisfy `(wd+pref)/R-tm0=0`, otherwise `TDS.init()` will fail.

Did this answer your question?

**Reply by xiaoheimai95** [reply]
When I change the equation of `Id`, `Iq`, and the logic of `PIcontroller`, the problem has been solved, however, it comes out with a new initialization problem.
The code of `Id` and `Iq` is as follows:
```
self.Id = Algeb(tex_name='I_d',
                info='d-axis current',
                v_str='Id0 + Kpvd * (vd + utd - vref2)',
                diag_eps=True,
                )
self.Iq = Algeb(tex_name='I_q',
                info='q-axis current',
                v_str='Iq0 + Kpvq * (vq + utq)',
                diag_eps=True,
                )
```
The code of the new `PIcontroller`:
```
self.xi.v_str = f'{self.x0.name}'
self.xi.e_str = f'{self.ki.name} * ({self.u.name} - {self.ref.name})'

self.y.v_str = f'{self.kp.name} * ({self.u.name} + {self.ut.name} - {self.ref.name}) + {self.x0.name}'
self.y.e_str = f'{self.kp.name} * ({self.u.name}  + {self.ut.name} - {self.ref.name}) + ' \
               f'{self.name}_xi - {self.name}_y'
```
And the problem is:
```
Suspect initialization issue! Simulation may crash!

    Name     | Var. Value | Eqn. Mismatch
-------------+------------+--------------
 Pe REGCV3 5 | 0.350      | 0.154
 Qe REGCV3 5 | 0.169      | -0.154

```

**Reply by xiaoheimai95** [reply]
The reason for this modification is that I only intend to add the control variable `ut` to the PI output part, then `Id` and `Iq` will also change because of the change of `PIvd_y` and `PIvq_y` here, not purely the original set initial values of `Id0` and `Iq0`.
My guess is that the `v_str` and `e_str` of `Pe` and `Qe` don't match anymore because of the change in `Id` and `Iq`, and I've observed the definitions of these two variables:
```
        self.Pe = Algeb(tex_name='P_e',
                        info='active power injection from VSC',
                        e_str='vd * Id + vq * Iq - Pe',
                        v_str='Pref')
        self.Qe = Algeb(tex_name='Q_e',
                        info='reactive power injection from VSC',
                        e_str='- vd * Iq + vq * Id - Qe',
                        v_str='Qref')
```
However the initial values of Pref and Qref I don't think should then be changed either. But I'm not so sure about my guesses and would like to see your opinion on it

**Reply by xiaoheimai95** [reply]
Is it possible to ensure consistency between `v_str` and `e_str` for `Pe` and `Qe` by changing the `v_str` in the code to the following format?
```
self.Pe.v_str = 'vd * (Id0 + Kpvd * (vd + utd - vref2)) + vq * (Iq0 + Kpvq * (vq + utq))'
self.Qe.v_str = '-vd * (Iq0 + Kpvq * (vq + utq)) + vq * (Id0 + Kpvd * (vd + utd - vref2))'
```

**Reply by xiaoheimai95** [reply]
Found out that this method will influence the value of `dw`, `a`, and `v`:
```
Suspect initialization issue! Simulation may crash!

    Name     | Var. Value | Eqn. Mismatch
-------------+------------+--------------
 dw REGCV3 5 | 0          | -0.154
 a Bus 8     | -0.022     | -0.154
 v Bus 8     | 1.030      | 0.155
```

**Comment by jinningwang** [comment]
Following comments for your consideration:
1. As the block ``PIControllerwithut`` is newly developed, we could focus on it firstly if initialization failed.
2. Since you developed a new block, two possible issues might occur: 1) the block definition itself has initialization issue, 2) the models that have this block has inappropriate initialization values.
3. 2.1 can be inspected by comparing with other existing block definitions, and their occurrence in models.
4. Once 2.1 is checked, 2.2 can be checked in the involved model.

As developing a new block can be a little bit complex, here is another easier way. You can skip developing a new block but implement the function you need in a model directly. It should be easier for initialization debug.

**Reply by xiaoheimai95** [reply]
As you mentioned, I modified the code of `VSGOuterPIModel` part as follows:
```
class VSGOuterPIModel:
    """
    Outer PI controllers for REGCV4 with external control variable ut.
    """

    def __init__(self, vderr: str = 'vd-vref2', vqerr: str = 'vq'):
        # PI controller for d-axis
        self.xid = State(info="d-axis PI controller integrator",
                         tex_name='x_{id}',
                         v_str='Id0 - Kpvd * (vd0 - vref2)',
                         e_str=f'Kivd * ({vderr}) * Qlim_zi')
        self.Idref = Algeb(info="d-axis current reference",
                           tex_name='I_{dref}',
                           v_str='Id0',
                           e_str=f'Kpvd * ({vderr} * Qlim_zi + xid + utd) - Idref')

        # PI controller for q-axis
        self.xiq = State(info="q-axis PI controller integrator",
                         tex_name='x_{iq}',
                         v_str='Iq0 - Kpvq * vq0',
                         e_str=f'Kivq * ({vqerr})')
        self.Iqref = Algeb(info="q-axis current reference",
                           tex_name='I_{qref}',
                           v_str='Iq0',
                           e_str=f'Kpvq * ({vqerr} + xiq + utq) - Iqref')
```
```
However, when I run TDS.init(), still got the result of Initialization failed:
Suspect initialization issue! Simulation may crash!

     Name       | Var. Value | Eqn. Mismatch
----------------+------------+--------------
 Idref REGCV4 3 | 0.396      | 0.100        
 Idref REGCV4 5 | 0.340      | 0.100        
 Iqref REGCV4 3 | -0.154     | 0.100        
 Iqref REGCV4 5 | -0.164     | 0.100        


Calculated TDS.h = 0.0333333
Initialization for dynamics completed in 0.0482 seconds.
Initialization failed!!
```
Also, because I wanted to impose some limitations on `Qe`, I added the `Qlim_zi` flag, which at this point has a value of 1 and doesn't affect what we're talking about.
